HTMLModelElement: fixes crash on model player creation during visibility change

amarchenko opened the pull request:
#### 026d4a3a9764a5db518f528108319dca7abdc9ce
<pre>
HTMLModelElement: fixes crash on model player creation during visibility change
<a href="https://bugs.webkit.org/show_bug.cgi?id=291540">https://bugs.webkit.org/show_bug.cgi?id=291540</a>
<a href="https://rdar.apple.com/149030547">rdar://149030547</a>

Reviewed by Ryosuke Niwa.

Checks if renderer is valid during contentSize() call, since there is
no guarantee for its availability. Also, subscribes for visibility
changes in insertedIntoAncestor() to make sure the element is notified
when in connected state. removedFromAncestor() is updated accordingly.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::m_environmentMapReadyPromise):
(WebCore::HTMLModelElement::~HTMLModelElement):
(WebCore::HTMLModelElement::contentSize const):
(WebCore::HTMLModelElement::insertedIntoAncestor):
(WebCore::HTMLModelElement::removedFromAncestor):

Canonical link: <a href="https://commits.webkit.org/293725@main">https://commits.webkit.org/293725@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a52eef67828dacc27f7dbd4eceb02982f226dc88

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99649 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9560 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104779 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50245 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/101690 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19661 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27732 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75854 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32952 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102656 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14980 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89980 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56213 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14780 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/7966 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49607 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84723 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8051 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107138 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26764 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19539 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84811 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27128 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86184 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84330 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21428 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29007 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6713 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/20554 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26704 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31966 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26519 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29835 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28089 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
```diff
@@ -1172,8 +1171,10 @@ Node::InsertedIntoAncestorResult HTMLModelElement::insertedIntoAncestor(Insertio
 
-    if (insertionType.connectedToDocument)
+    if (insertionType.connectedToDocument) {
+        document().registerForVisibilityStateChangedCallbacks(*this);
```
Now you can remove the calls to `registerForVisibilityStateChangedCallbacks` and `unregisterForVisibilityStateChangedCallbacks` in `HTMLModelElement::didMoveToNewDocument`.

amarchenko replied:
Done.

amarchenko committed:
HTMLModelElement: fixes crash on model player creation during visibility change
https://bugs.webkit.org/show_bug.cgi?id=291540
rdar://149030547

Reviewed by Ryosuke Niwa.

Checks if renderer is valid during contentSize() call, since there is
no guarantee for its availability. Also, subscribes for visibility
changes in insertedIntoAncestor() to make sure the element is notified
when in connected state. removedFromAncestor() is updated accordingly.

* Source/WebCore/Modules/model-element/HTMLModelElement.cpp:
(WebCore::m_environmentMapReadyPromise):
(WebCore::HTMLModelElement::~HTMLModelElement):
(WebCore::HTMLModelElement::contentSize const):
(WebCore::HTMLModelElement::insertedIntoAncestor):
(WebCore::HTMLModelElement::removedFromAncestor):

Canonical link: https://commits.webkit.org/293725@main
